# Exploring Ebay Car Sales Data

In [1]:
import pandas as pd
import numpy as np

autos= pd.read_csv('autos.csv',encoding='Latin-1')

In [2]:
autos.head()
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

From using autos.info we observed that:
1. Column vehicleType, gearbox,model,fueltype,notRepairDamage have null values.
2. Columns related to date does not have datetime object
3. Datatype of other columns should be converted as per data

In [3]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [4]:
autos.columns= ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We have converted column name from camelCase to Snake case as python prefer it

In [6]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

# Observing column Price and Odometer 

In [7]:
autos['price']= autos['price'].str.replace('$','').str.replace(",","").astype('int')

In [8]:
autos['odometer']= autos['odometer'].str.replace('km','').str.replace(",","").astype('int')

In [9]:
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

In [10]:
autos['odometer_km'].unique().shape

(13,)

In [11]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [12]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [13]:
autos['price']=autos.loc[autos['price'].between(1,350000),'price']

In [14]:
autos['price'].value_counts().sort_index(ascending=True)

1.0         156
2.0           3
3.0           1
5.0           2
8.0           1
9.0           1
10.0          7
11.0          2
12.0          3
13.0          2
14.0          1
15.0          2
17.0          3
18.0          1
20.0          4
25.0          5
29.0          1
30.0          7
35.0          1
40.0          6
45.0          4
47.0          1
49.0          4
50.0         49
55.0          2
59.0          1
60.0          9
65.0          5
66.0          1
70.0         10
           ... 
119500.0      1
119900.0      1
120000.0      2
128000.0      1
129000.0      1
130000.0      1
135000.0      1
137999.0      1
139997.0      1
145000.0      1
151990.0      1
155000.0      1
163500.0      1
163991.0      1
169000.0      1
169999.0      1
175000.0      1
180000.0      1
190000.0      1
194000.0      1
197000.0      1
198000.0      1
220000.0      1
250000.0      1
259000.0      1
265000.0      1
295000.0      1
299000.0      1
345000.0      1
350000.0      1
Name: price, Length: 234

- We observed that the column odometer have no outlier values
- In column price after 350000, the price becomes too large hence became outlier. So we have filter out and reassigned value of price upto 350000

In [15]:
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05 14:06:30    0.00002
2016-03-05 14:06:40    0.00002
2016-03-05 14:07:04    0.00002
2016-03-05 14:07:08    0.00002
2016-03-05 14:07:21    0.00002
2016-03-05 14:07:26    0.00002
2016-03-05 14:07:40    0.00002
2016-03-05 14:07:45    0.00002
2016-03-05 14:08:00    0.00004
2016-03-05 14:08:05    0.00004
2016-03-05 14:08:27    0.00002
2016-03-05 14:08:42    0.00002
2016-03-05 14:09:02    0.00004
2016-03-05 14:09:05    0.00002
2016-03-05 14:09:20    0.00002
2016-03-05 14:09:22    0.00002
2016-03-05 14:09:38    0.00002
2016-03-05 14:09:46    0.00002
2016-03-05 14:09:56    0.00002
2016-03-05 14:09:57    0.00002
2016-03-05 14:09:58    0.00004
2016-03-05 14:10:18    0.00002
2016-03-05 14:10:20    0.00002
2016-03-05 14:10:46    0.00002
2016-03-05 14:11:03    0.00002
2016-03-05 14:11:05    0.00002
2016-03-05 14:11:14    0.00002
2016-03-05 14:11:15    0.00002
2016-03-05 14:11:25    0.00002
2016-03-05 14:11:40    0.00002
                        ...   
2016-04-07 10:36:19    0.00002
2016-04-

In [16]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11 00:00:00    0.00002
2015-08-10 00:00:00    0.00002
2015-09-09 00:00:00    0.00002
2015-11-10 00:00:00    0.00002
2015-12-05 00:00:00    0.00002
2015-12-30 00:00:00    0.00002
2016-01-03 00:00:00    0.00002
2016-01-07 00:00:00    0.00002
2016-01-10 00:00:00    0.00004
2016-01-13 00:00:00    0.00002
2016-01-14 00:00:00    0.00002
2016-01-16 00:00:00    0.00002
2016-01-22 00:00:00    0.00002
2016-01-27 00:00:00    0.00006
2016-01-29 00:00:00    0.00002
2016-02-01 00:00:00    0.00002
2016-02-02 00:00:00    0.00004
2016-02-05 00:00:00    0.00004
2016-02-07 00:00:00    0.00002
2016-02-08 00:00:00    0.00002
2016-02-09 00:00:00    0.00004
2016-02-11 00:00:00    0.00002
2016-02-12 00:00:00    0.00006
2016-02-14 00:00:00    0.00004
2016-02-16 00:00:00    0.00002
2016-02-17 00:00:00    0.00002
2016-02-18 00:00:00    0.00004
2016-02-19 00:00:00    0.00006
2016-02-20 00:00:00    0.00004
2016-02-21 00:00:00    0.00006
                        ...   
2016-03-09 00:00:00    0.03324
2016-03-

In [17]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
2016-03-05 15:28:10    0.00002
2016-03-05 15:41:30    0.00002
2016-03-05 15:45:43    0.00002
2016-03-05 15:47:38    0.00002
2016-03-05 15:47:44    0.00002
2016-03-05 16:45:57    0.00002
2016-03-05 16:47:28    0.00002
2016-03-05 17:15:45    0.00002
2016-03-05 17:16:12    0.00002
2016-03-05 17:16:14    0.00002
2016-03-05 17:16:23    0.00002
2016-03-05 17:17:02    0.00002
2016-03-05 17:39:19    0.00002
2016-03-05 17:40:14    0.00002
2016-03-05 17:44:50    0.00002
2016-03-05 17:44:54    0.00002
2016-03-05 17:46:01    0.00002
2016-03-05 18:17:58    0.00002
2016-03-05 18:47:14    0.00002
2016-03-05 18:50:38    0.00002
2016-03-05 19:15:08    0.00002
2016-03-05 19:15:20    0.00002
2016-03-05 19:15:42    0.00002
2016-03-05 19:16:36    0.00002
2016-03-05 19:17:17    0.00002
                        ...   
2016-04-07 14:58:09    0.00004
2016-04-

In [18]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [19]:
autos['registration_year']=autos.loc[autos['registration_year'].between(1901,2020),'registration_year']

- We can see that min year is 1000 and max year is 9999. The min year is not absolutely correct as there would be no manufacturing back then so we have chosen 1901 as the min date.
- For max date we cannot predict what gonna happen in future. So we have chosen the current year (i.e 2020) as the max date.
- We have removed rows that have registration year below 1901 and above 2020

In [20]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

 1910.0    0.00018
 1927.0    0.00002
 1929.0    0.00002
 1931.0    0.00002
 1934.0    0.00004
 1937.0    0.00008
 1938.0    0.00002
 1939.0    0.00002
 1941.0    0.00004
 1943.0    0.00002
 1948.0    0.00002
 1950.0    0.00006
 1951.0    0.00004
 1952.0    0.00002
 1953.0    0.00002
 1954.0    0.00004
 1955.0    0.00004
 1956.0    0.00010
 1957.0    0.00004
 1958.0    0.00008
 1959.0    0.00014
 1960.0    0.00068
 1961.0    0.00012
 1962.0    0.00008
 1963.0    0.00018
 1964.0    0.00024
 1965.0    0.00034
 1966.0    0.00044
 1967.0    0.00054
 1968.0    0.00052
            ...   
 1991.0    0.00712
 1992.0    0.00782
 1993.0    0.00890
 1994.0    0.01320
 1995.0    0.02626
 1996.0    0.02888
 1997.0    0.04056
 1998.0    0.04906
 1999.0    0.06000
 2000.0    0.06708
 2001.0    0.05406
 2002.0    0.05066
 2003.0    0.05454
 2004.0    0.05474
 2005.0    0.06030
 2006.0    0.05416
 2007.0    0.04608
 2008.0    0.04462
 2009.0    0.04196
 2010.0    0.03194
 2011.0    0.03268
 2012.0    0

In [21]:
brands=autos['brand'].value_counts(normalize=True).index.tolist()[:20]

In [22]:
brand_agg= {}
for br in brands:
    mean_price= autos.loc[autos['brand']==br,'price'].mean()
    brand_agg[br]= mean_price
brand_agg

{'audi': 9212.9306621881,
 'bmw': 8261.382442169132,
 'citroen': 3761.8950437317785,
 'fiat': 2793.8700475435817,
 'ford': 3728.4121821407452,
 'hyundai': 5371.792960662526,
 'mazda': 4059.059539918809,
 'mercedes_benz': 8536.027085124677,
 'mini': 10541.566985645934,
 'nissan': 4669.3859649122805,
 'opel': 2944.6075421641085,
 'peugeot': 3065.611888111888,
 'renault': 2431.195698924731,
 'seat': 4315.744565217391,
 'skoda': 6353.544871794872,
 'smart': 3518.102305475504,
 'sonstige_autos': 12149.169851380042,
 'toyota': 5148.0032733224225,
 'volkswagen': 5332.4784249226,
 'volvo': 4866.993166287016}

In [23]:
brand_by_mileage= {}
for br in brands:
    mean_mileage= autos.loc[autos['brand']==br,'odometer_km'].mean()
    brand_by_mileage[br]= mean_mileage
brand_by_mileage

{'audi': 129643.9411627364,
 'bmw': 132521.64302818198,
 'citroen': 119764.62196861627,
 'fiat': 117037.4617737003,
 'ford': 124131.93446392642,
 'hyundai': 106782.7868852459,
 'mazda': 125132.10039630119,
 'mercedes_benz': 130886.14279678918,
 'mini': 89375.0,
 'nissan': 118978.7798408488,
 'opel': 129298.66324848929,
 'peugeot': 127352.33516483517,
 'renault': 128223.79367720465,
 'seat': 122061.63655685441,
 'skoda': 110947.83715012722,
 'smart': 100756.06276747503,
 'sonstige_autos': 87188.6446886447,
 'toyota': 115988.65478119935,
 'volkswagen': 128955.27276129878,
 'volvo': 138632.3851203501}

In [24]:
price_series= pd.Series(brand_agg)
print(price_series)

audi               9212.930662
bmw                8261.382442
citroen            3761.895044
fiat               2793.870048
ford               3728.412182
hyundai            5371.792961
mazda              4059.059540
mercedes_benz      8536.027085
mini              10541.566986
nissan             4669.385965
opel               2944.607542
peugeot            3065.611888
renault            2431.195699
seat               4315.744565
skoda              6353.544872
smart              3518.102305
sonstige_autos    12149.169851
toyota             5148.003273
volkswagen         5332.478425
volvo              4866.993166
dtype: float64


In [25]:
price_df= pd.DataFrame(price_series,columns=['mean_price'])
price_df

mean_price
audi             9212.930662
bmw              8261.382442
citroen          3761.895044
fiat             2793.870048
ford             3728.412182
hyundai          5371.792961
mazda            4059.059540
mercedes_benz    8536.027085
mini            10541.566986
nissan           4669.385965
opel             2944.607542
peugeot          3065.611888
renault          2431.195699
seat             4315.744565
skoda            6353.544872
smart            3518.102305
sonstige_autos  12149.169851
toyota           5148.003273
volkswagen       5332.478425
volvo            4866.993166

In [26]:
mileage_series= pd.Series(brand_by_mileage)
mileage_df= pd.DataFrame(mileage_series,columns=['mean_mileage'])
mileage_df

mean_mileage
audi            129643.941163
bmw             132521.643028
citroen         119764.621969
fiat            117037.461774
ford            124131.934464
hyundai         106782.786885
mazda           125132.100396
mercedes_benz   130886.142797
mini             89375.000000
nissan          118978.779841
opel            129298.663248
peugeot         127352.335165
renault         128223.793677
seat            122061.636557
skoda           110947.837150
smart           100756.062767
sonstige_autos   87188.644689
toyota          115988.654781
volkswagen      128955.272761
volvo           138632.385120

In [27]:
brand_df = pd.DataFrame(mileage_series,columns=['mean_mileage'])
brand_df

mean_mileage
audi            129643.941163
bmw             132521.643028
citroen         119764.621969
fiat            117037.461774
ford            124131.934464
hyundai         106782.786885
mazda           125132.100396
mercedes_benz   130886.142797
mini             89375.000000
nissan          118978.779841
opel            129298.663248
peugeot         127352.335165
renault         128223.793677
seat            122061.636557
skoda           110947.837150
smart           100756.062767
sonstige_autos   87188.644689
toyota          115988.654781
volkswagen      128955.272761
volvo           138632.385120

In [28]:
brand_df['mean_price']= price_series

In [29]:
brand_df

mean_mileage    mean_price
audi            129643.941163   9212.930662
bmw             132521.643028   8261.382442
citroen         119764.621969   3761.895044
fiat            117037.461774   2793.870048
ford            124131.934464   3728.412182
hyundai         106782.786885   5371.792961
mazda           125132.100396   4059.059540
mercedes_benz   130886.142797   8536.027085
mini             89375.000000  10541.566986
nissan          118978.779841   4669.385965
opel            129298.663248   2944.607542
peugeot         127352.335165   3065.611888
renault         128223.793677   2431.195699
seat            122061.636557   4315.744565
skoda           110947.837150   6353.544872
smart           100756.062767   3518.102305
sonstige_autos   87188.644689  12149.169851
toyota          115988.654781   5148.003273
volkswagen      128955.272761   5332.478425
volvo           138632.385120   4866.993166

Here we can conclude that mileage of costlier car is high